src/Collection.php

```diff
@@ -714,6 +716,11 @@ public function squash(): CollectionInterface
         return self::fromIterable($this->pack()->all())->unpack();
     }
 
+    public function strict(?callable $callback = null): CollectionInterface
+    {
+        return self::fromCallable(fn (): Generator => yield from (new TypedIterator($this->getIterator(), $callback)));
+    }
```

Comment:
I updated the PR with your gist and more tests, and added the `strict` operation now. I think this is nice, however, I'm wondering whether we could have an alternative so if I'm using the collection like a strict one I don't have to do `Collection::fromIterable([1,2])->strict()` all the time.

What if there was an optional boolean which was stored on a property on the class and based on it `getIterator` would use the Typed one or not. For example:


```PHP
public static function fromIterable(iterable $iterable, bool $strict = false): self
{
   	return new self(
        static fn (iterable $iterable): Iterator => new IterableIterator($iterable),
        $strict,
        $iterable
    );
}

public function getIterator(): Iterator
{
    $closureIterator = new ClosureIterator($this->source, ...$this->parameters);
	
    return $this->strict ? new TypedIterator($closureIterator): $closureIterator;
}


Collection::fromIterable([1, 2, 3], strict: true);
```

This would mean that the type enforcement is maintained throughout operation chaining, because of the way operations use the `getIterator` method.

Otherwise, the user would always have to remember to call `strict` after a chain of operations.

Does this sound like a crazy idea or does it make sense to you? 😄  @drupol

Reply:
I replied in the thread down there !